modules/odr_api/odr_api/api/endpoints/image.py

```diff
@@ -0,0 +1,184 @@
+import torch
+import numpy as np
+import torchvision.transforms as transforms
+import rawpy
+from fastapi import APIRouter, File, UploadFile, HTTPException
+from typing import Dict
+from io import BytesIO
+from PIL import Image, UnidentifiedImageError
+from PIL.ExifTags import TAGS
+from PIL.TiffImagePlugin import IFDRational
+import imageio
+from typing import Any
+import base64
+
+router = APIRouter(tags=["image"])
+
+
+# Helper functions for HDR stats
+def calculate_kurtosis(tensor: torch.Tensor):
+    # Calculate the mean and standard deviation
+    mean = torch.mean(tensor)
+    std = torch.std(tensor, unbiased=True)
+
+    # Calculate the fourth central moment
+    fourth_moment = torch.mean((tensor - mean) ** 4)
+
+    # Calculate kurtosis (excess kurtosis)
+    kurtosis_val = fourth_moment / (std ** 4)
+
+    return kurtosis_val - 3  # Subtract 3 for excess kurtosis (to make normal distribution kurtosis = 0)
+
+
+def calculate_msd(tensor: torch.Tensor):
+    mean = torch.mean(tensor)
+    msd = torch.mean((tensor - mean) ** 2)
+    return msd
+
+
+def calculate_dynamic_range(tensor: torch.Tensor, epsilon=1e-10):
+    I_min = torch.min(tensor)
+    I_max = torch.max(tensor)
+
+    # Prevent division by zero in both min and max
+    I_min = torch.clamp(I_min, min=epsilon)
+    I_max = torch.clamp(I_max, min=epsilon)
+
+    # Using decibels for DR
+    dynamic_range_db = 20 * torch.log10(I_max / I_min)
+    return dynamic_range_db
+
+
+def calculate_entropy(tensor: torch.Tensor):
+    # Flatten the tensor to 1D
+    tensor = tensor.flatten()
+
+    # Get unique values and their counts
+    unique_values, counts = torch.unique(tensor, return_counts=True)
+
+    # Calculate the probabilities of each unique value
+    probabilities = counts.float() / counts.sum()
+
+    # Calculate entropy using the Shannon entropy formula
+    entropy = -torch.sum(probabilities * torch.log2(probabilities))
+
+    return entropy
+
+
+# Helper functions for HDR metadata and preview conversion
+def extract_metadata(image_bytes: bytes, is_dng: bool) -> Dict:
```

Comment:
Have we tested a zipbomb style attack against this?

Reply:
No nothing like that yet, something we need to either protect to just allow through the front end, or to make more robust in an additional pass.